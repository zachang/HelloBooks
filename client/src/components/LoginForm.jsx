import React from 'react';

export default class LoginForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {}
  }

  render() {
    return (
      <form className="col s12">
        <div className="row">
          <div className="input-field col s12">
            <input id="email" type="email" className="validate"/>
            <label htmlFor="email">Email</label>
          </div>
        </div>

        <div className="row">
          <div className="input-field col s12">
            <input id="pass" type="password" className="validate"/>
            <label htmlFor="pass" data-error="wrong" data-success="right">Password</label>
          </div>
        </div>

        <div className="row">
          <div className="col s12">
            <p>
              <input type="checkbox" id="remember"/>
              <label htmlFor="remember">Remember me</label>
              <a href="#" className="right" style={{color: '#FF9800'}}>Forgot password</a>
            </p>
          </div>
        </div>

        <div className="row">
          <div className="col m12 s12">
            <button className="col m12 s12 btn btn-large waves-effect waves-light orange log-in"
                    type="button" name="action">Login
            </button>
          </div>
        </div>
      </form>
    );
  }
}
