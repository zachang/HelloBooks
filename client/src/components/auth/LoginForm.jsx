import React from 'react';
import PropTypes from 'react-proptypes';
import classnames from 'classnames';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import signinAction from '../../actions/signinAction.js';
import loginValidate from '../../utils/loginValidate.js';
import { redirectIfLoggedIn } from '../../utils/helpers';

export class LoginForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      loginCredentials: {
        username: '',
        password: ''
      },
      errors: {}
    };
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  componentWillMount() {
    if (window.sessionStorage.token) {
      redirectIfLoggedIn(window.sessionStorage.token);
    }
  }


  handleChange(e) {
    const loginCredentials = this.state.loginCredentials;
    loginCredentials[e.target.name] = e.target.value;
    this.setState({ loginCredentials })
  }

  isValid() {
    const { errors, isValid } = loginValidate(this.state.loginCredentials);
    if (!isValid) {
      this.setState({ errors });
    }
    return isValid;
  }

  handleSubmit(e) {
    e.preventDefault();
    if(this.isValid()) {
      this.setState({ errors: {} });
      this.props.signinAction(this.state.loginCredentials);
    }
  }

  render() {
    const { loginCredentials } = this.state;
    return (
      <form className='col s12' onSubmit={this.handleSubmit}>
        <div className='row'>
          <div className='input-field col s12'>
            <input
              id='username'
              name='username'
              type='text'
              className={ classnames({
                'invalid': !!this.state.errors.username
              })}
              value={loginCredentials.username}
              onChange={ this.handleChange }
            />
            <label
              htmlFor='username'
              className={(!!(this.state.errors['username']) || (loginCredentials.username.length > 0)) ?
                'custom-active custom-validate' : 'custom-validate'}
              data-error={ !!(this.state.errors['username']) ? this.state.errors['username'] : '' }
            >
              Username
            </label>
          </div>
        </div>

        <div className='row'>
          <div className='input-field col s12'>
            <input
              id='password'
              name='password'
              type='password'
              className={ classnames( {
                'invalid': !!this.state.errors.password
              })}
              value={loginCredentials.password}
              onChange={ this.handleChange }
            />
            <label
              htmlFor='password'
              className={ classnames( 'custom-validate', {
                'custom-active': (!!this.state.errors.password || (loginCredentials.password.length > 0))
              })}
              data-error={ !!(this.state.errors['password']) ? this.state.errors['password'] : '' }
            >
              Password
            </label>
          </div>
        </div>

        <div className='row'>
          <div className='col s12'>
            <p>
              <input type='checkbox' id='remember'/>
              <label htmlFor='remember'>Remember me</label>
            </p>
          </div>
        </div>

        <div className='row'>
          <div className='col m12 s12'>
            <button className='col m12 s12 btn btn-large waves-effect waves-light orange log-in'
              type='submit' name='action'>Login
            </button>

            <div style={{ color:'red', float:'right' }}>{this.props.signinState.fails}</div>
          </div>
        </div>
      </form>
    );
  }
}

LoginForm.propTypes = {
  signinState: PropTypes.object.isRequired,
  signinAction: PropTypes.func.isRequired
};
const mapStateToProps = state => ({
  signinState: state.signinReducer
});
const mapDispatchToProps = dispatch =>
  bindActionCreators({ signinAction }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(LoginForm);