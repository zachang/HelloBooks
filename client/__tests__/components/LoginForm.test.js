import React from 'react';
import toJson from 'enzyme-to-json';
import sinon from 'sinon';
import { shallow } from 'enzyme';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import storeConfiguration from '../../src/store/store';
import { LoginForm } from '../../src/components/auth/LoginForm.jsx';

const event = {
  preventDefault: jest.fn(),
  target: {
    username: 'pana8909',
    password: 'passphrase',
  }
};

const response = {
  profileObj: {
    email: 'pana8909@gmail.com',
    familyName: 'Maaten Stance',
    givenName: 'Winnie'
  }
};

describe('<LoginForm/>', () => {
  // create a spy function for signinAction
  const signinAction = sinon.spy();
  // create a spy function for googleSigninAction
  const googleSigninAction = sinon.spy();
  // create a spy function for redirectIfLoggedIn
  const redirectIfLoggedIn = sinon.spy();
  // spy on handleSubmit of Signin Page
  sinon.spy(LoginForm.prototype, 'handleSubmit');
  // spy on handleChange of Signin Page
  sinon.spy(LoginForm.prototype, 'handleChange');
  // spy on responseGoogle of Signin Page
  sinon.spy(LoginForm.prototype, 'isValid');
  // spy on responseGoogle of Signin Page
  sinon.spy(LoginForm.prototype, 'responseGoogle');
  // spy on componentWillMount of Signin Page
  sinon.spy(LoginForm.prototype, 'componentWillMount');
  // spy on componentWillReceiveProps of Signin Page
  sinon.spy(LoginForm.prototype, 'componentWillReceiveProps');

  const props = {
    signinState: {
      success: true,
      fails: null
    },
    signinAction,
    googleSigninAction,
  };

  const token = 'bbjsbjjsjhjnshugbjkgjg';
  const middleware = [thunk];
  const mockStore = configureMockStore(middleware);
  let store;
  let shallowComponent;
  beforeEach(() => {
    store = mockStore(storeConfiguration);
    shallowComponent = shallow(<LoginForm {...props} />);
    sessionStorage.setItem('token', token);
  });

  it('renders <LoginForm /> component', () => {
    const tree = toJson(shallowComponent);
    expect(tree).toMatchSnapshot();
  });

  it('should check that handleChange is called', () => {
    shallowComponent.instance().handleChange(event);
    expect(LoginForm.prototype.handleChange.calledOnce).toEqual(true);
  });

  it('should check that handleSubmit is called', () => {
    shallowComponent.instance().handleSubmit(event);
    expect(LoginForm.prototype.handleSubmit.calledOnce).toEqual(true);
  });

  it('calls componentWillReceiveProps', () => {
    props.signinState.success = false;
    shallowComponent.setState({
      fails: props.signinState.success
    });
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(LoginForm.prototype.componentWillReceiveProps.calledOnce).toEqual(true);
  });

  it('calls responseGoogle', () => {
    shallowComponent.instance().responseGoogle(response);
    expect(LoginForm.prototype.responseGoogle.calledOnce).toEqual(true);
    expect(googleSigninAction.called).toBe(true);
  });

  it('calls isValid', () => {
    shallowComponent.instance().isValid();
    expect(LoginForm.prototype.isValid.calledOnce).toEqual(true);
  });

  it('Should check that signinAction is called', () => {
    shallowComponent.setState({
      loginCredentials: {
        username: 'zenee09',
        password: 'kookoo',
      },
      errors: {},
      fails: null
    });
    shallowComponent.find('form').simulate('submit', { preventDefault() {} });
    expect(signinAction.called).toBe(true);
  });
});