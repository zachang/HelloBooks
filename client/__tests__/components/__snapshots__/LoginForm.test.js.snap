// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm/> renders <LoginForm /> component 1`] = `
<form
  className="col s12"
  onSubmit={[Function]}
>
  <div
    className="row"
  >
    <div
      className="input-field col s12"
    >
      <input
        className=""
        id="username"
        name="username"
        onChange={[Function]}
        type="text"
        value=""
      />
      <label
        className="custom-validate"
        data-error=""
        htmlFor="username"
      >
        Username
      </label>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="input-field col s12"
    >
      <input
        className=""
        id="password"
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
      <label
        className="custom-validate"
        data-error=""
        htmlFor="password"
      >
        Password
      </label>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col s12"
    >
      <p>
        <input
          id="remember"
          type="checkbox"
        />
        <label
          htmlFor="remember"
        >
          Remember me
        </label>
      </p>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col m12 s12"
    >
      <div
        className="row"
      >
        <button
          className="col m12 s12 btn btn-large waves-effect waves-light orange log-in"
          name="action"
          type="submit"
        >
          Login
        </button>
      </div>
      <div
        className="row"
      >
        <t
          buttonText="Google"
          className="col m12 s12 btn btn-large waves-effect waves-light log-in"
          cookiePolicy="single_host_origin"
          disabledStyle={
            Object {
              "opacity": 0.6,
            }
          }
          fetchBasicProfile={true}
          isSignedIn={false}
          onFailure={[Function]}
          onRequest={[Function]}
          onSuccess={[Function]}
          prompt=""
          scope="profile email"
          tag="button"
          type="button"
          uxMode="popup"
        >
          <SocialIcon
            className="google"
            color="white"
            network="google"
          />
          <span>
             Login
          </span>
        </t>
      </div>
      <div
        style={
          Object {
            "color": "red",
            "float": "right",
          }
        }
      />
    </div>
  </div>
</form>
`;
